external-providers/java-external-provider/pkg/java_external_provider/service_client.go

```diff
@@ -184,22 +184,79 @@ func (p *javaServiceClient) GetAllSymbols(ctx context.Context, c javaCondition,
 		}
 	}
 
-	if c.Referenced.Filepaths != nil {
-		// filter according to the given filepaths
-		var filteredRefs []protocol.WorkspaceSymbol
-		for _, ref := range refs {
-			for _, fp := range c.Referenced.Filepaths {
-				if strings.HasSuffix(ref.Location.Value.(protocol.Location).URI, fp) {
-					filteredRefs = append(filteredRefs, ref)
-				}
-			}
+	// Use FileSearcher to handle filepath filtering from multiple sources:
+	// 1. Rule scope excluded paths (language server only knows about included paths)
+	// 2. Condition filepaths (with proper pattern matching and normalization)
+	// The java language server indexes both source and dependencies (e.g., Maven repo).
+	// The language server was already instructed to filter by provider-level and rule-scope
+	// included paths. Here we apply additional filtering for excluded paths and
+	// condition-level filepaths.
+	includedFilepaths, excludedFilepaths := condCTX.GetScopedFilepaths()
+
+	// Check if we have any additional filtering to apply beyond what the language server did
+	// The language server filters at package-level, we need file-level precision
+	hasAdditionalConstraints := len(includedFilepaths) > 0 || len(excludedFilepaths) > 0 || len(c.Referenced.Filepaths) > 0
+
+	if !hasAdditionalConstraints {
+		// No additional constraints - return language server results as-is
+		return refs, nil
+	}
+
+	// Determine base path - use workspace directory
+	basePath := p.workspace
+
+	fileSearcher := provider.FileSearcher{
+		BasePath: basePath,
+		ProviderConfigConstraints: provider.IncludeExcludeConstraints{
+			IncludePathsOrPatterns: p.includedPaths,
+		},
+		RuleScopeConstraints: provider.IncludeExcludeConstraints{
+			IncludePathsOrPatterns: includedFilepaths,
+			ExcludePathsOrPatterns: excludedFilepaths,
+		},
+		FailFast: true,
+		Log:      log,
+	}
+
+	// Search for allowed files
+	paths, err := fileSearcher.Search(provider.SearchCriteria{
+		ConditionFilepaths: c.Referenced.Filepaths,
+	})
```

Comment:
nit: we can do this in a follow up

We should create and start searching before doing the other work and wait for serach to finish before we filter out. Much like we are doing for the nodejs provider.